### ISSUE TITLE
Extracting assets

### ISSUE THREAD

**Issue by merak69** [DerPopo/UABE#82]
So after spending the better part of a day reading through the deepest parts of the internet and playing around with your API, I'm at a loss... Does the API support extraction of assets to be saved out en masse?

I see how to open an asset file and I can see from other examples online how to replace existing assets with new ones and write out the assets file again.  I don't see how to get access to the asset itself after creating AssetsFile and AssetsFileTable to dump that stream to its own file.

I've combed through the header files that currently exist for the API and went as far using dumpbin to see that there are definitely way more things not exposed in the current header files.

My prior code worked up until the UnityFS version came out and the game I work with has moved up to 5.3.x.

Any help or direction you can provide would be greatly appreciated, if you're willing.  Thank you!

**Comment by merak69** [comment]
I went back and read post by post of the forum thread and figured out what I was overlooking...though the API could use some documentation :)

I got to the point where I can open the bundle, extract the assets files and then I was working through getting the texture2D objects I want however I ran into a problem and I'm unsure what's going on.

According to the AssetsFileTable, the asset I want to process is at offset 4208 in the file.  I setup a ClassDatabase as follows:

`	AssetTypeTemplateField* pBaseField = new AssetTypeTemplateField;
	ClassDatabaseFile* classDatabaseFile = new ClassDatabaseFile();
	classDatabaseFile->Read(AssetsReaderFromFile, (LPARAM)pClassFile);
	pBaseField->FromClassDatabase(classDatabaseFile, &classDatabaseFile->classes[28], (DWORD)0);`


However, as soon as I do the following line, it hangs and never returns.

`AssetTypeInstance ati = AssetTypeInstance(1, &pBaseField, AssetsReaderFromFile, (LPARAM)pInputFile, 4208);`

Am I just overlooking something dumb?

**Comment by DerPopo** [comment]
The indices of ClassDatabaseFile::classes aren't the same as the type/class id. You'll need to go through the classes list and find the one with classId == 28.

**Comment by merak69** [comment]
Thanks for the comment DerPopo.  I happened to notice that last night when I was getting Skybox as the type.  I switched it to 22 (Texture2D) and then realized I was loading one version too old of the ClassDatabaseFile.  Fixing both of those resolved the hang.

I then proceeded to do the ReadTextureFile call which worked great and it filled out the TextureFile object.  When I tried to use the GetTextureData method, the pOutBut always came out as zero bytes in size for an input format of PVRTC_RGB2.

When I use "TextureFile"->pPictureData, I get the raw data obviously minus the PVR headers.  I added them in and was able to output the PVR file which I could then convert but I'm not sure why the GetTextureData never worked.  Does it only support certain formats today?
